### DISCUSSION TITLE
Empty router query on initial render

### DISCUSSION THREAD

**Discussion by LunarMelody** [vercel/next.js#39184] [Help]
Hello! I have encountered quite annoying problem of empty query object on initial render on router. I do know that it's intended and I have also seen that there's a big discussion about it. However, I have also discovered that we can technically extract query from `router.asPath`, but I am not sure how safe and/or good it is :thinking: I'd love to hear some advises! 

I came up with a hook like this: 

```typescript
import { useRouter } from "next/router";
import { parse } from "qs";

export const useRouterQuery = () => {
  const { asPath } = useRouter();

  const urlAndParams = asPath.split("?", 2);
  const params = urlAndParams.at(1) ?? "";

  return parse(params);
};
```

P.S. Is there actally a better way of getting query without this workaround? Thanks!

**Comment by GabenGar** [comment]
`asPath` will not populate queries from the dynamic path segments. Also you can go even further and just feed the `asPath` to `URL` constructor and get `URL.searchParams` from it, which is an idiomatic way to work with search params in modern JS.

**Reply by LunarMelody** [reply]
That didn't really work unfortunately, passing `asPath` to `URL` throws an invalid URL error :smile: To be fair, the goal I am trying to achieve doesn't even really work properly unfortunately. The hook itself kinda doesn't work too, when it's on a dev server the hook works just fine but fails in production with hydration mismatches, guess I'll have to use initial props or maybe there's a better workaround I just don't really know about :( 

To be fair I wish router was a little bit more flexible in Next, I suppose I'll have to wait for a rewrite version I've been hearing about :thinking:
